source/doc/settingsFile.rst
.. _settings-page:

Settings File
=================================


|projname| stores its settings in the JSON_ format. When using the graphical interface, a settings file is generated automatically upon starting an analysis, or by clicking *Save Settings* in the *File menu*  (see :ref:`gui-page`).

Settings Layout
---------------------------------------------

JSON_ is a human readable file format that consists of key-value pairs separated by sections. Each section in a JSON object consists of a section name and a list of string key-value pairs. 

.. sourcecode:: javascript

	{
		"<section name>" : {
			"key1" : "value1",
			"key2" : "value2",
			...
		}

	}

|projname| settings define a new section for each class, with key-value pairs corresponding to class attributes that are set upon initialization. This is illustrated below for the stepResponseAnalysis class. The *stepResponseAnalysis* section in the settings file holds parameters corresponding to the :class:`~mosaic.stepResponseAnalysis` class. Note that that the section name in the settings file is identical to the corresponding class name. Three parameters are then defined within the section that control the behavior of the class.

.. sourcecode:: javascript

	{
		"stepResponseAnalysis" : {
			"FitTol"			: "1.e-7",
			"FitIters"			: "50000",
			"BlockRejectRatio"		: "0.9"
		}
	}

Finally, :class:`~mosaic.stepResponseAnalysis` is initialized by defining class attributes corresponding to the key-value pairs in the settings file.

.. sourcecode:: python
	
	try:
			self.FitTol=float(self.settingsDict.pop("FitTol", 1.e-7))
			self.FitIters=int(self.settingsDict.pop("FitIters", 5000))

			self.BlockRejectRatio=float(self.settingsDict.pop("BlockRejectRatio", 0.8))
		
	except ValueError as err:
		raise commonExceptions.SettingsTypeError( err )


Trajectory Settings
---------------------------------------------

Common Settings (:class:`~mosaic.metaTrajIO.metaTrajIO`)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. exec::
   import mosaic.metaTrajIO
   
   print mosaic.metaTrajIO.metaTrajIO.__doc__


QDF Files (:class:`~mosaic.qdfTrajIO.qdfTrajIO`)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. exec::
   import mosaic.qdfTrajIO
   
   print mosaic.qdfTrajIO.qdfTrajIO.__doc__


ABF Files (:class:`~mosaic.abfTrajIO.abfTrajIO`)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. exec::
   import mosaic.abfTrajIO
   
   print mosaic.abfTrajIO.abfTrajIO.__doc__

Binary Files (:class:`~mosaic.binTrajIO.binTrajIO`)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. exec::
   import mosaic.binTrajIO
   
   print mosaic.binTrajIO.binTrajIO.__doc__

.. _algorithm-settings-sec:

Optimizing Settings
---------------------------------------------

|projname| classes are controlled through the JSON_ settings files as defined above. In most cases, running |projname| through the GUI (see :ref:`gui-page`) should generate satisfactory results. However, settings can be further optimized either by editing a file named ``.settings`` stored within the data directory, or by clicking on the ``Advanced Settings`` check-box in the :ref:`Analysis-Setup` section of the GUI.

.. _eventseg-settings-sec:

Initial Event Detection (:class:`~mosaic.eventSegment.eventSegment`)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The first step when analyzing an ionic-current time series is to perform a quick partition to identify events. This is accomplished by overriding the :class:`~mosaic.eventPartition.eventPartition` class. Currently, the only implementation of event partitioning is the :class:`~mosaic.eventSegment.eventSegment` algorithm. This algorithm uses a thresholding technique to detect the start and end of an event. When an event is detected the ionic current time-series associated with that event is passed to a processing algorithm for fitting. Settings that can be passed to :class:`~mosaic.eventSegment.eventSegment` are given below followed by their descriptions.


.. sourcecode:: javascript


	"eventSegment" : {
		"blockSizeSec" 			: "0.5",
		"eventPad" 			: "50",
		"minEventLength" 		: "5",
		"eventThreshold" 		: "6.0",
		"driftThreshold" 		: "999.0",
		"maxDriftRate" 			: "999.0",
		"meanOpenCurr"			: "-1",
		"sdOpenCurr"			: "-1",
		"slopeOpenCurr"			: "-1",
		"writeEventTS"			: "1",
		"parallelProc"			: "0",
		"reserveNCPU"			: "2"
	}


.. tabularcolumns:: p{4cm}p{12cm}

+-------------------+----------------------------------------------------------------------------------------+
|  **Setting**      | **Description**                                                                        |
+===================+========================================================================================+
| blockSizeSec      | Time-series length (in sec) for block operations.                                      |
|                   |                                                                                        |
| eventPad          | Pad an event with the specified number of points.                                      |
|                   |                                                                                        | 
| minEventLength    | Discard events with fewer than the specfied points.                                    |
|                   |                                                                                        | 
| eventThreshold    | Event detection threshold.                                                             |
|                   |                                                                                        | 
| meanOpenCurr      | Set the mean open channel current (i0) in pA. -1 computes i0 automatically.            |
|                   |                                                                                        | 
| sdOpenCurr        | Set the open channel std. dev. in pA. -1 computes SD automatically.                    |
|                   |                                                                                        | 
| slopeOpenCurr     | Set the open channel drift in pA/ms. -1 automatically computes the slope.              |
|                   |                                                                                        | 
| driftThreshold    | Aborts the analysis when the open channel drift exceeds the specified value.           |
|                   |                                                                                        | 
| maxDriftRate      | Aborts the analysis when the open channel slope exceeds the specified value (pA/ms).   |
|                   |                                                                                        | 
| writeEventTS      | Write the event time-series to the output database.                                    |
|                   |                                                                                        | 
| parallelProc      | Enable parallel processing.                                                            |
|                   |                                                                                        | 
| reserveNCPU       | Use N-reserveNCPU for parallel processing.                                             |
|                   |                                                                                        | 
+-------------------+----------------------------------------------------------------------------------------+


.. _stepresp-settings-sec:

Two-State Identification  (:class:`~mosaic.stepResponseAnalysis.stepResponseAnalysis`)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Once the time-series is partitioned, individual events are processed by a processing algorithm. For simple event patterns (e.g. homopolymers of DNA, PEG, etc.), one can use the :ref:`stepresponse-page` algorithm. Settings that can be passed to this algorithm are below, followed by their descriptions. For a vast majority of cases, the settings below can be used without modification.

.. sourcecode:: javascript

	"stepResponseAnalysis" : {
		"FitTol"			: "1.e-7",
		"FitIters"			: "50000"
	}

.. tabularcolumns:: p{4cm}p{12cm}

+-------------------+----------------------------------------------------------------------------------------+
|  **Setting**      | **Description**                                                                        |
+===================+========================================================================================+
| FitTol            | Controls the convergence criteria for the least squares fitting algorithm.             |
|                   |                                                                                        |
| FitIters          | Maximum iterations passed to the least squares fitting algorithm.                      |
|                   |                                                                                        | 
+-------------------+----------------------------------------------------------------------------------------|  


Multi-State Identification  (:class:`~mosaic.multiStateAnalysis.multiStateAnalysis`)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more complex signals with multiple states, the :ref:`multistate-page` algorithm yields better results. The settings passed to this algorithm (described below) are largely similar to :ref:`stepresp-settings-sec`. 

.. sourcecode:: javascript
	:emphasize-lines: 4

	"multiStateAnalysis" : {
		"FitTol"			: "1.e-7",
		"FitIters"			: "50000",
		"InitThreshold"			: "3.0"
	}

.. tabularcolumns:: p{4cm}p{12cm}

+-------------------+----------------------------------------------------------------------------------------+
|  **Setting**      | **Description**                                                                        |
+===================+========================================================================================+
| FitTol            | Controls the convergence criteria for the least squares fitting algorithm.             |
|                   |                                                                                        |
| FitIters          | Maximum iterations passed to the least squares fitting algorithm.                      |
|                   |                                                                                        | 
| InitThreshold     | Threshold used to identify the states in a multi-state event.                          |
|                   |                                                                                        | 
+-------------------+----------------------------------------------------------------------------------------|  

.. hint:: The parameter ``InitThreshold`` is used for preliminary state identification within multi-state events. As a rule of thumb, this value should be set to roughly half that of ``eventThreshold`` in the :ref:`eventseg-settings-sec` section. However, the final value may be adjusted further for optimal results.

Default Settings
---------------------------------------------

.. sourcecode:: javascript

	{
		"eventSegment" : {
			"blockSizeSec" 			: "0.5",
			"eventPad" 				: "50",
			"minEventLength" 		: "5",
			"eventThreshold" 		: "6.0",
			"driftThreshold" 		: "999.0",
			"maxDriftRate" 			: "999.0",
			"meanOpenCurr"			: "-1",
			"sdOpenCurr"			: "-1",
			"slopeOpenCurr"			: "-1",
			"writeEventTS"			: "1",
			"parallelProc"			: "0",
			"reserveNCPU"			: "2"
		},
		"singleStepEvent" : {
			"binSize" 				: "1.0",
			"histPad" 				: "10",
			"maxFitIters"			: "5000",
			"a12Ratio" 				: "1.e4",
			"minEvntTime" 			: "10.e-6",
			"minDataPad" 			: "75"
		},
		"stepResponseAnalysis" : {
			"FitTol"				: "1.e-7",
			"FitIters"				: "50000"
		},
		"multiStateAnalysis" : {
	                "FitTol"		: "1.e-7",
	                "FitIters"		: "50000",
	                "InitThreshold"	: "3.0"
	     },
	     "cusumLevelAnalysis": {
        			"StepSize"		: 3.0, 
        			"Threshold"		: 3.0
    	}, 
		"besselLowpassFilter" : {
			"filterOrder"			: "6",
			"filterCutoff"			: "10000",
			"decimate"				: "1"	
		},
		"waveletDenoiseFilter" : {
			"wavelet"				: "sym5",
			"level"					: "5",
			"thresholdType"			: "soft",
			"thresholdSubType"		: "sqtwolog"
		},
		"abfTrajIO" : {
			"filter"				: "*.abf", 
			"start"					: 0.0, 
			"dcOffset"				: 0.0
		},
		"qdfTrajIO": {
			"Rfb": 9.1e+12, 
			"Cfb": 1.07e-12, 
			"dcOffset": 0.0, 
			"filter": "*.qdf", 
			"start": 0.0
		},
		"binTrajIO": {
			"AmplifierScale": "1.0", 
			"AmplifierOffset": "0.0", 
			"SamplingFrequency": "50000",
			"HeaderOffset": "0",
			"ColumnTypes": "[('curr_pA', 'float64')]",
			"IonicCurrentColumn" : "curr_pA",
			"dcOffset": "0.0", 
			"filter": "*.bin", 
			"start": "0.0"
		}
	}


.. include:: ../aliases.rst



doc-sphinx/doc/settingsFile.rst
.. _settings-page:

Settings File
=================================


|projname| stores its settings in the |json| format. When using the graphical interface, a settings file is generated automatically upon starting an analysis, or by clicking *Save Settings* in the *File menu*  (see :ref:`gui-page`).

Settings Layout
---------------------------------------------

|json| is a human readable file format that consists of key-value pairs separated by sections. Each section in a JSON object consists of a section  name and a list of string key-value pairs. 

.. sourcecode:: javascript

	{
		"<section name>" : {
			"key1" : "value1",
			"key2" : "value2",
			...
		}

	}

|projname| settings define a new section for each class, with key-value pairs corresponding to class attributes set upon initialization. This is illustrated below for the stepResponseAnalysis class. The settings file defines a section corresponding to the class name *stepResponseAnalysis*. Three parameters are then defined within the section that control the behavior of the class.

.. sourcecode:: javascript

	{
		"stepResponseAnalysis" : {
			"FitTol"			: "1.e-7",
			"FitIters"			: "50000",
			"BlockRejectRatio"		: "0.9"
		}
	}

The code below from the initialization of *stepResponseAnalysis* in :class:`pyeventanalysis.stepResponseAnalysis` sets three class attributes corresponding to the key-value pairs in the settings file.

.. sourcecode:: python
	
	try:
			self.FitTol=float(self.settingsDict.pop("FitTol", 1.e-7))
			self.FitIters=int(self.settingsDict.pop("FitIters", 5000))

			self.BlockRejectRatio=float(self.settingsDict.pop("BlockRejectRatio", 0.8))
		
	except ValueError as err:
		raise commonExceptions.SettingsTypeError( err )



Default Settings
---------------------------------------------

.. sourcecode:: javascript

	{
		"eventSegment" : {
			"blockSizeSec" 			: "0.5",
			"eventPad" 			: "50",
			"minEventLength" 		: "5",
			"eventThreshold" 		: "6.0",
			"driftThreshold" 		: "999.0",
			"maxDriftRate" 			: "999.0",
			"meanOpenCurr"			: "-1",
			"sdOpenCurr"			: "-1",
			"slopeOpenCurr"			: "-1",
			"writeEventTS"			: "1",
			"parallelProc"			: "0",
			"reserveNCPU"			: "2",
			"plotResults"			: "0"
		},
		"singleStepEvent" : {
			"binSize" 			: "1.0",
			"histPad" 			: "10",
			"maxFitIters"			: "5000",
			"a12Ratio" 			: "1.e4",
			"minEvntTime" 			: "10.e-6",
			"minDataPad" 			: "75"
		},
		"stepResponseAnalysis" : {
			"FitTol"			: "1.e-7",
			"FitIters"			: "50000",
			"BlockRejectRatio"		: "0.9"
		},
		"multiStateAnalysis" : {
			"FitTol"			: "1.e-7",
			"FitIters"			: "50000",
			"InitThreshold"			: "5.0"
		},
		"besselLowpassFilter" : {
			"filterOrder"			: "6",
			"filterCutoff"			: "10000",
			"decimate"			: "1"	
		},
		"waveletDenoiseFilter" : {
			"wavelet"			: "sym5",
			"level"				: "5",
			"thresholdType"			: "soft",
			"thresholdSubType"		: "sqtwolog"
		}
	}

mosaic/settings.py
"""
	Load analysis settings from a JSON file. 

	:Created:	8/24/2012
 	:Author: 	Arvind Balijepalli <arvind.balijepalli@nist.gov>
	:License:	See LICENSE.TXT
	:ChangeLog:
	.. line-block::
		2/14/15 	AB 	Added default settings for cusumLevelAnalysis.
		8/20/14		AB 	Changed precedence of settings file search to datpath/.settings,
						datpath/settings, coderoot/.settings and coderoot/settings
		8/6/14		AB 	Add a function to parse a settings string.
		9/5/13		AB 	Check for either .settings or settings in data directory
						and code root. Warn when using default settings
		8/24/12		AB	Initial version	
"""
import json
import os
import os.path

class settings:
	"""
			Initialize a settings object. 

			:Args: 
				- `datpath` :	Specify the location of the settings file. If a settings file is not found, return default settings.
				- `defaultwarn` :	If `True` warn the user if a settings file was not found in the path specified by `datpath`.
	"""
	def __init__(self, datpath, defaultwarn=True):
		"""
		"""	
		self.settingsFile=None

		if os.path.isfile(datpath+'/.settings'):
			self.settingsFile=datpath+"/.settings"
			settingstr="".join((open(self.settingsFile, 'r').readlines()))
		elif os.path.isfile(datpath+'/settings'):
			self.settingsFile=datpath+"/settings"
			settingstr="".join((open(self.settingsFile, 'r').readlines()))
		# elif os.path.isfile('.settings'):
		# 	print "Settings file not found in data directory. Default settings will be used."
		# 	self.settingsFile=os.getcwd()+"/.settings"
		# elif os.path.isfile('settings'):
		# 	print "Settings file not found in data directory. Default settings will be used."
		# 	self.settingsFile=os.getcwd()+"/settings"
		else:
			if defaultwarn:
				print "Settings file not found in data directory. Default settings will be used."
			settingstr=__settings__


		self.parseSettingsString( settingstr )

	def parseSettingsString(self, settingstring):
		self.settingsDict=json.loads( settingstring )

	def getSettings(self, section):
		"""
			Return settings for a specified section as a Python dict.

			:Args:
				- `section` : 	specifies the section for which settings are requested. Returns an empty dictionary if the settings file doesn't exist the section is not found.
		"""
		try:
			return self.settingsDict[section]
		except KeyError, AttributeError:
			return {}

__settings__="""
	{
		"eventSegment" : {
			"blockSizeSec" 			: "0.5",
			"eventPad" 				: "50",
			"minEventLength" 		: "5",
			"eventThreshold" 		: "6.0",
			"driftThreshold" 		: "999.0",
			"maxDriftRate" 			: "999.0",
			"meanOpenCurr"			: "-1",
			"sdOpenCurr"			: "-1",
			"slopeOpenCurr"			: "-1",
			"writeEventTS"			: "1",
			"parallelProc"			: "0",
			"reserveNCPU"			: "2"
		},
		"singleStepEvent" : {
			"binSize" 				: "1.0",
			"histPad" 				: "10",
			"maxFitIters"			: "5000",
			"a12Ratio" 				: "1.e4",
			"minEvntTime" 			: "10.e-6",
			"minDataPad" 			: "75"
		},
		"stepResponseAnalysis" : {
			"FitTol"				: "1.e-7",
			"FitIters"				: "50000"
		},
		"multiStateAnalysis" : {
	                "FitTol"		: "1.e-7",
	                "FitIters"		: "50000",
	                "InitThreshold"	: "3.0"
	     },
	     "cusumLevelAnalysis": {
        			"StepSize"		: 3.0, 
        			"Threshold"		: 3.0
    	}, 
		"besselLowpassFilter" : {
			"filterOrder"			: "6",
			"filterCutoff"			: "10000",
			"decimate"				: "1"	
		},
		"waveletDenoiseFilter" : {
			"wavelet"				: "sym5",
			"level"					: "5",
			"thresholdType"			: "soft",
			"thresholdSubType"		: "sqtwolog"
		},
		"abfTrajIO" : {
			"filter"				: "*.abf", 
			"start"					: 0.0, 
			"dcOffset"				: 0.0
		},
		"qdfTrajIO": {
			"Rfb": 9.1e+12, 
			"Cfb": 1.07e-12, 
			"dcOffset": 0.0, 
			"filter": "*.qdf", 
			"start": 0.0
		},
		"binTrajIO": {
			"AmplifierScale": "1.0", 
			"AmplifierOffset": "0.0", 
			"SamplingFrequency": "50000",
			"HeaderOffset": "0",
			"ColumnTypes": "[('curr_pA', 'float64')]",
			"IonicCurrentColumn" : "curr_pA",
			"dcOffset": "0.0", 
			"filter": "*.bin", 
			"start": "0.0"
		}
	}
"""